pageserver/src/tenant/block_io.rs

```diff
@@ -43,37 +42,34 @@ where
     }
 }
 
-/// A block accessible for reading
-///
-/// During builds with `#[cfg(test)]`, this is a proper enum
-/// with two variants to support testing code. During normal
-/// builds, it just has one variant and is thus a cheap newtype
-/// wrapper of [`PageReadGuard`]
-pub enum BlockLease {
+/// Reference to an in-memory copy of an immutable on-disk block.
+pub enum BlockLease<'a> {
     PageReadGuard(PageReadGuard<'static>),
+    EphemeralFileMutableTail(&'a [u8; PAGE_SZ]),
     #[cfg(test)]
     Rc(std::rc::Rc<[u8; PAGE_SZ]>),
 }
 
-impl From<PageReadGuard<'static>> for BlockLease {
+impl<'a> From<PageReadGuard<'static>> for BlockLease<'a> {
```

Comment:
This does look weird, can we get non-`'static` out of this? But the lifetime is variant specific, this cannot be anything other than `BlockLease<'static>`, right?

Reply:
Yeah, I can change the signature like so if that makes you happy/happier:

```
-    fn from(value: PageReadGuard<'static>) -> Self {
+    fn from(value: PageReadGuard<'static>) -> BlockLease<'static> {
```

Reply:
I'm bothered by the `'a` where only possibility is `'static` but that would go at the same time.

Reply:
I think at the moment with both `'a` and `'static` it makes the least sense. I am surprised if this even compiles because ... well, I guess the same substitution still applies, `'a = 'static`. But the special thing with `'static` is that we don't need or want to be generic over the single one `'static`.

Reply:
Not sure if you noticed, but, the `'a` is for the MutableTail variant, not for the page cache guard variant.

Reply:
Yes, but here the `'a` cannot be anything but `'static`. So the issue still exists.

```suggestion
impl From<PageReadGuard<'static>> for BlockLease<'static> {
```

This is what I am after (after your `from(value: PageReadGuard<'static>) -> BlockLease<'static>` change)

Reply:
> the 'a cannot be anything but 'static

Technically, `'a` *can* be a non-static lifetime as well. The `'a` is chosen by the user of `BlockLease`. In the end it doesn't matter that much though because `BlockLease<'static>` coerces to any other `BlockLease<'a>`. @koivunej 's proposed solution is shorter so that's good I guess.

Reply:
Yes, the substitution is allowed, but this just confusing way to express it because the impl block is generic over lifetime but then the it can only be one.

Reply:
Applied suggestion in 8e891e3bd666f90c8fea2b58febeb594704ab362